### ISSUE TITLE
helm push with error when subchart using apiVerion v2

### ISSUE THREAD

**Issue by LittleWhiteYA** [chartmuseum/helm-push#93]
## Summary

Hi !

I got a strange error when I'm using `helm push`

```
$ helm push mychart/ chartmuseum --debug
Error: error unpacking subchart-x.x.x.tgz in mychart: apiVersion 'v2' is not valid. The value must be "v1"

Usage:
  helm push [flags]

Flags:
      --access-token string             Send token in Authorization header [$HELM_REPO_ACCESS_TOKEN]
      --auth-header string              Alternative header to use for token auth [$HELM_REPO_AUTH_HEADER]
...

Error: plugin "push" exited with error
helm.go:94: [debug] plugin "push" exited with error
```

I think maybe can not use `apiVersion: v2` in chart, but I digged into subchart and try to do the same thing using `helm push` to push subchart into chartmuseum like belong:

```
$ helm push subchart/ chartmuseum
```

Nothing error happens ...

I googled a few hours, but can not find related issue.

Hope someone could give me some directions 

## version

helm push version: 
```
$ helm plugin list
NAME       	VERSION	DESCRIPTION                                                                              
push       	0.9.0  	Push chart package to ChartMuseum                                         
```

helm version:
```
$ helm version
version.BuildInfo{Version:"v3.3.4", GitCommit:"a61ce5633af99708171414353ed49547cf05013d", GitTreeState:"clean", GoVersion:"go1.14.9"}
```

**Comment by jdolitsky** [comment]
@LittleWhiteYA do you still get the error if you first package the chart? and push a .tgz file?

**Comment by samrees** [comment]
same error if you package the chart:

```
bash-5.0# helm push canary.tgz chartmuseum
Error: error unpacking redis in canary: apiVersion 'v2' is not valid. The value must be "v1"
Usage:
  helm push [flags]

```

in this example, canary.tgz is a v1 chart, depending on a v2 chart.

**Comment by karuppiah7890** [comment]
@LittleWhiteYA @samrees I just found the issue. I tried these two cases -

- v1 chart depending on v2 chart with subchart as tar ball in `charts` directory
- v1 chart depending on v2 chart with subchart as requirements.yaml and using `-d` option for dependency update

There is issue in both cases. Reason - when trying to load a api v1 chart, helm push uses Helm v2 library, and in this case, there is no support for loading api v2 sub charts (Helm v3) in api v1 charts. This is inherently a feature that is not present in Helm v2.

And no such feature will be added in the future. But I can start checking the Helm issues just to see if someone has spoken about this and what was the conclusion. 

What I would recommend is - can you check if you can use api v2 charts with Helm v3? if that's not possible, then we gotta see what to do.

But I'm afraid that the further push plugin support for api v1, that is Helm v2, is kind of unnecessary at this point as the official project itself has stopped support - https://helm.sh/blog/helm-v2-deprecation-timeline/ , https://github.com/helm/helm/releases/tag/v2.17.0 . Especially when the feature has to be made on the helm package (ideally).

Do we want to support Helm v2 charts with Helm v3 subcharts? If yes, we need to spend time on writing code similar to Helm package to support Helm v2 charts (api v1) with Helm v3 sub charts

If not, I think that this issue can be closed 😅

cc @jdolitsky

**Comment by karuppiah7890** [comment]
GitHub search of issues is not helping much with respect to this issue, so I'm just gonna let others to checkout the same in helm repo. You can check https://github.com/helm/charts/ repo too maybe :)

**Comment by karuppiah7890** [comment]
I started digging deep into this.

One thing to note is, I'm still able to package Helm v2 charts with Helm v3 sub charts and also push using `curl` or some similar HTTP client.

I do wish that this can be tried to be solved differently. I'm checking more. But the thing is, with loading the chart and subcharts, we check if the directory / tar ball is actually a chart, and also, we get the feature to override the version of the chart. Apart from that, we don't do much by loading the chart. We finally package the chart and push it

**Comment by karuppiah7890** [comment]
Looks like the code causing the issue has been added in this PR https://github.com/chartmuseum/helm-push/pull/53 

I was thinking of other ways to solve the issue. Maybe we might just be able to get it to work :) Since the packaging of the files was the problem in #52 , what we can do is, we can always load with Helm v3 library, which can load both api v1 (Helm v2) and api v2 (Helm v3) charts. Only while packaging, we can convert the Helm v3 Chart struct to Helm v2 chart in case the chart is a Helm v2 chart.

I'm going try if this solution works :D

**Comment by karuppiah7890** [comment]
Another tricky thing is - the chart could be like this too -

Helm v2 chart > Helm v3 sub chart > Helm v2 sub chart > Helm v3 sub chart > ...

So, we need to ensure that the appropriate charts have the appropriate files and don't lose any files like #52 mentions

**Comment by karuppiah7890** [comment]
@jdolitsky I have reverted the golang code changes made in 17794c3f46c90a7518814cdac5bb834c3adc9708  . Looks like it isn't necessary now. At the time of #52 , in v0.8.0 , the helm v3 library used was v3.0.1 which didn't have code to help with loading and saving helm v2 charts. But that's not the case anymore with newer versions of helm v3 library. So, I have reverted any golang code changes we made assuming that helm v3 cannot help with loading and saving/packaging charts. The acceptances tests written for the issue still pass, to show that it all still works. 

PR is at #96
